src/include/storage/record_buffer.h

```diff
@@ -169,19 +297,53 @@ class UndoBuffer {
    * @param size the size of the undo record to allocate
    * @return a new undo record with at least the given size reserved
    */
-  UndoRecord *NewEntry(const uint32_t size) {
-    if (buffers_.empty() || !buffers_.back()->HasBytesLeft(size)) {
-      // we are out of space in the buffer. Get a new buffer segment.
-      BufferSegment *new_segment = buffer_pool_->Get();
-      TERRIER_ASSERT(reinterpret_cast<uintptr_t>(new_segment) % 8 == 0, "a delta entry should be aligned to 8 bytes");
-      new_segment->Reset();
-      buffers_.push_back(new_segment);
-    }
-    return reinterpret_cast<UndoRecord *>(buffers_.back()->Reserve(size));
-  }
+  byte *NewEntry(uint32_t size);
 
  private:
-  common::ObjectPool<BufferSegment> *buffer_pool_;
+  RecordBufferSegmentPool *buffer_pool_;
   std::vector<BufferSegment *> buffers_;
 };
+
+class LogManager;  // forward declaration
+
+/**
+ * A RedoBuffer is a fixed-sized buffer to hold RedoRecords.
+ *
+ * Not thread-safe
+ */
+class RedoBuffer {
+ public:
+  /**
+   * Initializes a new RedoBuffer, working with the given LogManager
+   * @param log_manager the log manager this redo buffer talks to, or nullptr if logging is disabled
+   * @param buffer_pool The buffer pool to draw buffer segments from. Must be the same buffer pool the log manager uses.
+   */
+  explicit RedoBuffer(LogManager *log_manager, RecordBufferSegmentPool *buffer_pool)
+      : log_manager_(log_manager), buffer_pool_(buffer_pool) {}
+
+  /**
+   * Reserve a redo record with the given size. The returned pointer is guaranteed to be valid until NewEntry is called
+   * again, or when the buffer is explicitly flushed by the call Finish().
+   * @param size the size of the redo record to allocate
```

Comment:
The documentation is kind of confusing since we use RedoBuffer to allocate both RedoRecord and CommitRecord. Maybe we should rename RedoBuffer to be something like LogBuffer?? But since it's a very large PR, we can do refactoring in a separate PR